treebuilder.ts

```diff
@@ -0,0 +1,160 @@
+import path from 'path';
+import * as ts from 'typescript';
+import glob from 'glob';
+import cmd from 'node-cmd';
+
+type Node = {
+  fileName: string;
+  dependsOn: Node[];
+  dependedOnBy: Node[];
+};
+
+type GraphType = {
+  [name: string]: Node;
+};
+
+const graph: GraphType = {};
+
+function compile(fileNames: string[], options: ts.CompilerOptions): void {
+  const program = ts.createProgram(fileNames, options);
+
+  fileNames.map((fileName) => {
+    if (!graph[path.resolve(fileName)]) {
+      const ast: any = program.getSourceFile(fileName);
+      recurse({
+        fileName: skipIndexFile(path.resolve(ast.originalFileName)),
+        dependsOn: [],
+        dependedOnBy: [],
+      });
+    }
+  });
+
+  function recurse(node: Node) {
+    const ast: any = program.getSourceFile(node.fileName);
+
+    if (ast && ast.resolvedModules) {
+      const dependencies = Array.from(ast.resolvedModules.entries())
+        .map(([key, module]: any) => {
+          if (!module) {
+            return recurse({
+              fileName: path.resolve(
+                ast.originalFileName.replace(/(.*\/)(.*)/, `$1${key}`),
+              ),
+              dependsOn: [],
+              dependedOnBy: [node],
+            });
+          }
+
+          if (module.isExternalLibraryImport) {
+            return undefined;
+          }
+
+          const moduleFileName = module.resolvedFileName;
+
+          let newNode;
+          if (graph[path.resolve(moduleFileName)]) {
+            newNode = graph[path.resolve(moduleFileName)];
+            newNode.dependedOnBy.push(node);
+          } else {
+            newNode = recurse({
+              fileName: skipIndexFile(moduleFileName),
+              dependsOn: [],
+              dependedOnBy: [node],
+            });
+          }
+          return newNode;
+        })
+        .filter((node) => node);
+
+      node.dependsOn = dependencies;
+    }
+
+    graph[path.resolve(node.fileName)] = node;
+    return node;
+  }
+
+  function skipIndexFile(fileName) {
+    if (/(components\/)(\w*\/)?(index.ts)/.test(fileName)) {
```

Comment:
@amrocha This seems to cause mischief in some cases, can you have a look?

### Steps to reproduce

1. Make a change to `src/components/Modal/components/Footer/Footer.tsx`
2. Run `yarn observer`

### Expected

```
🧩  src/components/Modal/components/Footer/Footer.tsx
    Component name         File potentially affected (total: 5)
    <Frame />              src/components/Frame/Frame.tsx
    <Frame />              src/components/Frame/components/ContextualSaveBar/ContextualSaveBar.tsx
    <Frame />              src/components/Frame/components/ContextualSaveBar/components/DiscardConfirmationModal/DiscardConfirmationModal.tsx
    <Modal />              src/components/Modal/Modal.tsx
    <Toast />              src/components/Toast/Toast.tsx
```

### Actual

```
🧩  src/components/Modal/components/Footer/Footer.tsx
    Component name         File potentially affected (total: 0)
```

Reply:
Oh interesting, taking a look